libp2p-identify/src/transport.rs

```diff
@@ -0,0 +1,384 @@
+					let real_addr = match identify {
+						IdentifyOutput::RemoteInfo { info, .. } => process_identify_info(
+							&info,
+							&*peerstore.clone(),
+							original_addr,
+							addr_ttl,
+						)?,
+						_ => unreachable!(
+							"the identify protocol guarantees that we receive \
+							 remote information when we dial a node"
+						),
+					};
+
+					Ok((connec, real_addr))
+				});
+
+			Box::new(fut) as Box<Future<Item = _, Error = _>>
+		});
+
+		Ok((Box::new(listener) as Box<_>, new_addr))
+	}
+
+	#[inline]
+	fn dial(self, addr: Multiaddr) -> Result<Self::Dial, (Self, Multiaddr)> {
+		match multiaddr_to_peerid(addr.clone()) {
+			Ok(peer_id) => {
+				// If the multiaddress is a peer ID, try each known multiaddress (taken from the
+				// peerstore) one by one.
+				let addrs = self.peerstore
+					.deref()
+					.peer(&peer_id)
+					.into_iter()
+					.flat_map(|peer| peer.addrs())
+					.collect::<Vec<_>>()
+					.into_iter();
+
+				let transport = self.transport;
+				let future = stream::iter_ok(addrs)
+					// Try to dial each address through the transport.
+					.filter_map(move |addr| transport.clone().dial(addr).ok())
+					.and_then(move |dial| dial)
+					// Pick the first non-failing dial result.
+					.then(|res| Ok(res))
+					.filter_map(|res| res.ok())
+					.into_future()
+					.map_err(|(err, _)| err)
+					.and_then(|(val, _)| val.ok_or(IoErrorKind::InvalidData.into())) // TODO: wrong error
+					.map(move |(socket, _inner_client_addr)| (socket, addr));
+
+				Ok(Box::new(future) as Box<_>)
+			}
+
+			Err(addr) => {
+				// If the multiaddress is something else, propagate it to the underlying transport
+				// and identify the node.
+				let transport = self.transport;
+				let identify_upgrade = transport.clone().with_upgrade(IdentifyProtocolConfig);
+
+				// We dial a first time the node and upgrade it to identify.
+				let dial = match identify_upgrade.dial(addr) {
+					Ok(d) => d,
+					Err((_, addr)) => {
+						let id = IdentifyTransport {
+							transport,
+							peerstore: self.peerstore,
+							addr_ttl: self.addr_ttl,
+						};
+						return Err((id, addr));
+					}
+				};
+
+				let peerstore = self.peerstore;
+				let addr_ttl = self.addr_ttl;
+
+				let future = dial.and_then(move |identify| {
+					// On success, store the information in the peerstore and compute the
+					// "real" address of the node (of the form `/p2p/...`).
+					let (real_addr, old_addr);
+					match identify {
+						(IdentifyOutput::RemoteInfo { info, .. }, a) => {
+							old_addr = a.clone();
+							real_addr = process_identify_info(&info, &*peerstore, a, addr_ttl)?;
+						}
+						_ => unreachable!(
+							"the identify protocol guarantees that we receive \
+							 remote information when we dial a node"
+						),
+					};
+
+					// Then dial the same node again.
+					Ok(transport
+						.dial(old_addr)
+						.unwrap_or_else(|_| {
+							panic!("the same multiaddr was determined to be valid earlier")
```

Comment:
Maybe this panic/message is a sign that we should have a way to dial using an already-parsed `Multiaddr`, since this is always infallible if the multiaddr is already verified. It'd be nice to express that in the API instead of relying on the user to know that.

Reply:
On the other hand, I don't see any other situation where we would dial the same multiaddress twice in a row, and adding such an API would add a lot of complexity.

Reply:
What about the use-case of compile-time checked constant multiaddrs that are parsed by a macro? I think enforcing strings to be the canonical representation of multiaddrs at API boundaries is a poor choice and we should probably have a trait akin to `ToSocketAddrs` (it would be possible to have the `dial` call be infallible for types that can infallibly be converted to a `Multiaddr` using type hackery but I'd prefer if we just added an associated `Error` type to this trait and made `type Error = !` when that lands). I don't think that's necessary for this PR though, I'll write up another PR with that in it.